export const SOURCE_RAW_SEQUENCING_DATA_KEYS = [
  "coverage_100kb_plus",
  "1Mb+",
  "200kb+",
  "300kb+",
  "400kb+",
  "500kb+",
  "DeepConsensus_version",
  "Family ID",
  "Gender",
  "MM_remove",
  "MM_review",
  "MM_tag",
  "Maternal ID",
  "N25",
  "N50",
  "N75",
  "Other Comments",
  "Paternal ID",
  "Phenotype",
  "Population",
  "Relationship",
  "Second Order",
  "Siblings",
  "Third Order",
  "accession",
  "barcode",
  "basecaller",
  "basecaller_model",
  "basecaller_version",
  "bioproject_accession",
  "biosample_accession",
  "ccs_algorithm",
  "cell_type",
  "check-flnc reads",
  "coverage",
  "data_type",
  "deepconsensus_coverage",
  "deepconsensus_filename",
  "deepconsensus_path",
  "design_description",
  "filename",
  "filetype",
  "generator_contact",
  "generator_facility",
  "instrument_model",
  "iso_filename",
  "iso_library_id",
  "jasmine_version",
  "library_ID",
  "library_construction_protocol",
  "library_layout",
  "library_selection",
  "library_source",
  "library_strategy",
  "lima_float_version",
  "lima_version",
  "max",
  "mean",
  "min",
  "notes",
  "ntsm_score",
  "path",
  "pbtrim_version",
  "phasing",
  "platform",
  "platform_unit_1",
  "platform_unit_2",
  "polymerase_version",
  "pool",
  "primrose_filename",
  "production",
  "quartile_25",
  "quartile_50",
  "quartile_75",
  "read_N50",
  "read_length",
  "refine_version",
  "sample_ID",
  "seq_kit",
  "seq_plate_chemistry_version",
  "sequencing_chemistry",
  "shear_method",
  "similarity",
  "size_selection",
  "study",
  "title",
  "total_Gbp",
  "total_bp",
  "total_reads",
  "whales",
  "sample_id",
  "biosample_id",
  "population_descriptor",
  "population_abbreviation",
  "trio_available",
  "family_id",
  "paternal_id",
  "maternal_id",
  "sex",
  "tissue",
  "collection",
  "alternative_id",
  "file_size",
] as const;

export const SOURCE_ASSEMBLY_KEYS = [
  "sample_id",
  "haplotype",
  "assembly",
  "fasta_sha256",
  "release",
  "phasing",
  "assembly_method",
  "assembly_method_version",
  "assembly_date",
  "assembly_name",
  "source",
  "genbank_accession",
  "assembly_md5",
  "assembly_fai",
  "assembly_gzi",
  "biosample_id",
  "population_descriptor",
  "population_abbreviation",
  "trio_available",
  "family_id",
  "paternal_id",
  "maternal_id",
  "sex",
  "tissue",
  "collection",
  "alternative_id",
  "notes",
  "file_size",
] as const;

export const SOURCE_ANNOTATION_KEYS = [
  "sample_id",
  "haplotype",
  "location",
  "annotation_type",
  "release",
  "assembly_name",
  "file_size",
] as const;

export const SOURCE_ALIGNMENT_KEYS = [
  "alignment",
  "file",
  "file_size",
  "loc",
  "pipeline",
  "reference_coordinates",
  "use_case",
  "version",
] as const;

export const HAPLOTYPE_BY_ID: Record<string, string> = {
  1: "paternal",
  2: "maternal",
};
